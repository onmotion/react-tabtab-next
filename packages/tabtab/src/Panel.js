"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PanelStyle = void 0;
var React = __importStar(require("react"));
var styled_components_1 = __importDefault(require("styled-components"));
var PanelStyle = styled_components_1.default.div(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  background-color: white;\n  text-align: left;\n  padding: 20px 15px;\n  ", "\n"], ["\n  background-color: white;\n  text-align: left;\n  padding: 20px 15px;\n  ", "\n"])), function (props) { return !props.active ? "display: none;" : null; });
exports.PanelStyle = PanelStyle;
var PanelComponent = (function (_super) {
    __extends(PanelComponent, _super);
    function PanelComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    PanelComponent.prototype.render = function () {
        var _a = this.props, active = _a.active, index = _a.index;
        var Panel = this.props.CustomPanelStyle || PanelStyle;
        return (React.createElement(Panel, { role: "tabpanel", id: "react-tabtab-panel-".concat(index), "aria-labelledby": "react-tabtab-".concat(index), "aria-hidden": false, active: active }, active ?
            this.props.children
            : null));
    };
    return PanelComponent;
}(React.PureComponent));
exports.default = PanelComponent;
var templateObject_1;
