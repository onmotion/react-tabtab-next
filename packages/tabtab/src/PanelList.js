"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = __importStar(require("react"));
var PanelList = (function (_super) {
    __extends(PanelList, _super);
    function PanelList() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    PanelList.prototype.render = function () {
        var _a = this.props, children = _a.children, activeIndex = _a.activeIndex, customStyle = _a.customStyle;
        if (!children || activeIndex === undefined) {
            return null;
        }
        var props = {};
        if (customStyle && customStyle.Panel) {
            props = __assign(__assign({}, props), { CustomPanelStyle: customStyle.Panel });
        }
        var result = React.Children.toArray(children).map(function (child, index) { return (React.cloneElement(child, __assign({ key: index, active: index === activeIndex, index: index }, props))); });
        return React.createElement("div", null, result);
    };
    return PanelList;
}(React.PureComponent));
exports.default = PanelList;
