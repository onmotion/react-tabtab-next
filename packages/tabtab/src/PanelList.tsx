// @flow
import * as React from 'react';
import { ReactElement } from 'react';

type Props = {
    activeIndex?: number;
    customStyle?: {
        Panel: () => void;
    };
};

export default class PanelList extends React.PureComponent<React.PropsWithChildren<Props>> {
    render() {
        const { children, activeIndex, customStyle } = this.props;
        if (!children || activeIndex === undefined) {
            return null;
        }

        let props = {};
        if (customStyle && customStyle.Panel) {
            props = { ...props, CustomPanelStyle: customStyle.Panel };
        }

        // to prevent the type of one children is object type
        const result = React.Children.toArray(children).map((child, index) =>
            React.cloneElement(child as ReactElement, {
                key: index,
                active: index === activeIndex,
                index,
                ...props,
            })
        );
        return <div>{result}</div>;
    }
}
