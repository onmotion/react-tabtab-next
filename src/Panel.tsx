// @flow
import * as React from 'react'
import styled from 'styled-components';

const PanelStyle = styled.div`
  background-color: white;
  text-align: left;
  padding: 20px 15px;
  ${props => !props.active ? `display: none;` : null}
`;

type Props = {
  children: React.ReactNode,
  CustomPanelStyle: () => void,
  active: boolean,
  index: number
};

export default class PanelComponent extends React.PureComponent<Props> {
  render() {
    const {active, index} = this.props;
    const Panel = this.props.CustomPanelStyle || PanelStyle;
    return (
      <Panel role="tabpanel"
             id={`react-tabtab-panel-${index}`}
             aria-labelledby={`react-tabtab-${index}`}
             aria-hidden={false}
             active={active}>
        {active ?
          this.props.children
        : null}
      </Panel>
    )
  }
}

export {
  PanelStyle
}