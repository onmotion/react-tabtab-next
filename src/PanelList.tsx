// @flow
import * as React from 'react';

type Props = {
  children: Array<any>,
  activeIndex: number,
  customStyle: {
    Panel: () => void
  }
};

export default class PanelList extends React.PureComponent<Props> {
  render() {
    const {
      children,
      activeIndex,
      customStyle
    } = this.props;
    if (!children || activeIndex === undefined) {
      return null;
    }

    let props = {};
    if (customStyle && customStyle.Panel) {
      props = {...props, CustomPanelStyle: customStyle.Panel}
    }

    // to prevent the type of one children is object type
    const result = React.Children.toArray(children).map((child, index) => (
      React.cloneElement(child, {
        key: index,
        active: index === activeIndex,
        index,
        ...props
      })
    ));
    return <div>{result}</div>
  }
}
