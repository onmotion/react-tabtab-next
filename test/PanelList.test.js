import React from 'react';
import PanelList from '../src/PanelList';
import Panel, {PanelStyle} from '../src/Panel';
import {shallow, mount} from 'enzyme';
import toJson from 'enzyme-to-json';
import 'jest-styled-components';

describe('render', () => {
  test('render one Panel', () => {
    const component = mount(
      <PanelList activeIndex={0}>
        <Panel>
          <span>panel content</span>
        </Panel>
      </PanelList>
    );
    expect(toJson(component)).toMatchSnapshot();
  });

  test('render multi Panel', () => {
    const component = mount(
      <PanelList activeIndex={1}>
        <Panel>
          <span>panel content</span>
        </Panel>
        <Panel>
          <span>panel content</span>
        </Panel>
        <Panel>
          <span>panel content</span>
        </Panel>
      </PanelList>
    );
    expect(toJson(component)).toMatchSnapshot();    
  });

  test('return null if no child', () => {
    const component = shallow(
      <PanelList>
      </PanelList>
    );
    expect(component.html()).toEqual(null);
  })
})

test('custom style', () => {
  const component = shallow(
    <PanelList activeIndex={1}
               customStyle={{
                 Panel: PanelStyle
               }}>
      <Panel>
        <span>panel content</span>
      </Panel>
    </PanelList>
  );
  expect(toJson(component)).toMatchSnapshot();
})