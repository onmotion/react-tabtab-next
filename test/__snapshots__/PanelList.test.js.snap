// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`custom style 1`] = `
<div>
  <PanelComponent
    CustomPanelStyle={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": undefined,
        "componentStyle": ComponentStyle {
          "componentId": "sc-bdVaJa",
          "isStatic": false,
          "lastClassName": "guOOuH",
          "rules": Array [
            "
        background-color: white;
        text-align: left;
        padding: 20px 15px;
        ",
            [Function],
            "
      ",
          ],
        },
        "displayName": "styled.div",
        "render": [Function],
        "styledComponentId": "sc-bdVaJa",
        "target": "div",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      }
    }
    active={false}
    index={0}
    key="0"
  >
    <span>
      panel content
    </span>
  </PanelComponent>
</div>
`;

exports[`render render multi Panel 1`] = `
.c1 {
  background-color: white;
  text-align: left;
  padding: 20px 15px;
}

.c0 {
  background-color: white;
  text-align: left;
  padding: 20px 15px;
  display: none;
}

<PanelList
  activeIndex={1}
>
  <div>
    <PanelComponent
      active={false}
      index={0}
      key="0"
    >
      <styled.div
        active={false}
        aria-hidden={false}
        aria-labelledby="react-tabtab-0"
        id="react-tabtab-panel-0"
        role="tabpanel"
      >
        <StyledComponent
          active={false}
          aria-hidden={false}
          aria-labelledby="react-tabtab-0"
          forwardedClass={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": undefined,
              "componentStyle": ComponentStyle {
                "componentId": "sc-bdVaJa",
                "isStatic": false,
                "lastClassName": "c0",
                "rules": Array [
                  "
              background-color: white;
              text-align: left;
              padding: 20px 15px;
              ",
                  [Function],
                  "
            ",
                ],
              },
              "displayName": "styled.div",
              "render": [Function],
              "styledComponentId": "sc-bdVaJa",
              "target": "div",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          forwardedRef={null}
          id="react-tabtab-panel-0"
          role="tabpanel"
        >
          <div
            aria-hidden={false}
            aria-labelledby="react-tabtab-0"
            className="c0"
            id="react-tabtab-panel-0"
            role="tabpanel"
          />
        </StyledComponent>
      </styled.div>
    </PanelComponent>
    <PanelComponent
      active={true}
      index={1}
      key="1"
    >
      <styled.div
        active={true}
        aria-hidden={false}
        aria-labelledby="react-tabtab-1"
        id="react-tabtab-panel-1"
        role="tabpanel"
      >
        <StyledComponent
          active={true}
          aria-hidden={false}
          aria-labelledby="react-tabtab-1"
          forwardedClass={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": undefined,
              "componentStyle": ComponentStyle {
                "componentId": "sc-bdVaJa",
                "isStatic": false,
                "lastClassName": "c0",
                "rules": Array [
                  "
              background-color: white;
              text-align: left;
              padding: 20px 15px;
              ",
                  [Function],
                  "
            ",
                ],
              },
              "displayName": "styled.div",
              "render": [Function],
              "styledComponentId": "sc-bdVaJa",
              "target": "div",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          forwardedRef={null}
          id="react-tabtab-panel-1"
          role="tabpanel"
        >
          <div
            aria-hidden={false}
            aria-labelledby="react-tabtab-1"
            className="c1"
            id="react-tabtab-panel-1"
            role="tabpanel"
          >
            <span>
              panel content
            </span>
          </div>
        </StyledComponent>
      </styled.div>
    </PanelComponent>
    <PanelComponent
      active={false}
      index={2}
      key="2"
    >
      <styled.div
        active={false}
        aria-hidden={false}
        aria-labelledby="react-tabtab-2"
        id="react-tabtab-panel-2"
        role="tabpanel"
      >
        <StyledComponent
          active={false}
          aria-hidden={false}
          aria-labelledby="react-tabtab-2"
          forwardedClass={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": undefined,
              "componentStyle": ComponentStyle {
                "componentId": "sc-bdVaJa",
                "isStatic": false,
                "lastClassName": "c0",
                "rules": Array [
                  "
              background-color: white;
              text-align: left;
              padding: 20px 15px;
              ",
                  [Function],
                  "
            ",
                ],
              },
              "displayName": "styled.div",
              "render": [Function],
              "styledComponentId": "sc-bdVaJa",
              "target": "div",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          forwardedRef={null}
          id="react-tabtab-panel-2"
          role="tabpanel"
        >
          <div
            aria-hidden={false}
            aria-labelledby="react-tabtab-2"
            className="c0"
            id="react-tabtab-panel-2"
            role="tabpanel"
          />
        </StyledComponent>
      </styled.div>
    </PanelComponent>
  </div>
</PanelList>
`;

exports[`render render one Panel 1`] = `
.c0 {
  background-color: white;
  text-align: left;
  padding: 20px 15px;
}

<PanelList
  activeIndex={0}
>
  <div>
    <PanelComponent
      active={true}
      index={0}
      key="0"
    >
      <styled.div
        active={true}
        aria-hidden={false}
        aria-labelledby="react-tabtab-0"
        id="react-tabtab-panel-0"
        role="tabpanel"
      >
        <StyledComponent
          active={true}
          aria-hidden={false}
          aria-labelledby="react-tabtab-0"
          forwardedClass={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": undefined,
              "componentStyle": ComponentStyle {
                "componentId": "sc-bdVaJa",
                "isStatic": false,
                "lastClassName": "c0",
                "rules": Array [
                  "
              background-color: white;
              text-align: left;
              padding: 20px 15px;
              ",
                  [Function],
                  "
            ",
                ],
              },
              "displayName": "styled.div",
              "render": [Function],
              "styledComponentId": "sc-bdVaJa",
              "target": "div",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          forwardedRef={null}
          id="react-tabtab-panel-0"
          role="tabpanel"
        >
          <div
            aria-hidden={false}
            aria-labelledby="react-tabtab-0"
            className="c0"
            id="react-tabtab-panel-0"
            role="tabpanel"
          >
            <span>
              panel content
            </span>
          </div>
        </StyledComponent>
      </styled.div>
    </PanelComponent>
  </div>
</PanelList>
`;
